File: adjuster-portal-ui/src/app/config/advance-schedule.ts
/**
 * ACI Commission Engine — Advance Schedule.
 *
 * Single source of truth for the tier/cap policy that governs team-member
 * cash advances against in-progress claims. Consumed by IssueAdvanceDialog
 * (computes the tier amount + enforces caps) and, eventually, by any batch
 * advance-eligibility job that rolls forward at month-end.
 *
 * Keep this file in sync with the authoritative policy doc. Mutating
 * tiers here changes what the UI offers and what the server validates —
 * make changes deliberately and communicate them to ops.
 */

export interface AdvanceTier {
  /** Minimum estimate (inclusive) that qualifies for this tier. */
  min: number;
  /**
   * Maximum estimate (inclusive) for this tier, or `null` for the open-ended
   * top bracket. An estimate E matches a tier when `min <= E && (max === null || E <= max)`.
   */
  max: number | null;
  /** The advance amount (USD) granted for claims in this tier. */
  amount: number;
}

export interface AdvanceSchedule {
  /**
   * Ordered list of tiers. `computeTierAmount` assumes increasing-min.
   */
  tiers: readonly AdvanceTier[];
  /**
   * Hard cap on the sum of ADVANCE_ISSUED amounts within a single Monday-
   * to-Sunday calendar week, per team member.
   */
  weeklyCapPerMember: number;
  /**
   * Hard cap on the lifetime sum of ADVANCE_ISSUED amounts per team member.
   */
  lifetimeCapPerMember: number;
  /**
   * Policy token for estimates below the lowest tier minimum. Only meaningful
   * value today is `"admin_discretionary"` — the dialog hides the tier
   * amount, surfaces a free-text amount input, and requires the admin
   * override toggle before the submit button enables.
   */
  underMinimumPolicy: 'admin_discretionary';
}

export const ADVANCE_SCHEDULE: AdvanceSchedule = {
  tiers: [
    { min: 10_000,  max: 50_000,  amount: 250 },
    { min: 50_001,  max: 100_000, amount: 500 },
    { min: 100_001, max: 200_000, amount: 1_000 },
    { min: 200_001, max: null,    amount: 1_500 },
  ],
  weeklyCapPerMember: 5_000,
  lifetimeCapPerMember: 25_000,
  underMinimumPolicy: 'admin_discretionary',
};

/**
 * Returns the tier-matched advance amount for a given claim estimate, or
 * null if the estimate falls below the lowest tier's minimum (in which
 * case the dialog routes to admin-discretionary free-text input).
 */
export function computeTierAmount(
  estimate: number | null | undefined,
  schedule: AdvanceSchedule = ADVANCE_SCHEDULE,
): number | null {
  if (estimate == null || estimate <= 0) return null;
  for (const tier of schedule.tiers) {
    if (estimate >= tier.min && (tier.max === null || estimate <= tier.max)) {
      return tier.amount;
    }
  }
  // Estimate falls below the lowest tier.min → admin-discretionary.
  return null;
}

File: upa-portal/backend/app/app/config/advance_schedule.py
#!/usr/bin/env python

"""ACI Commission Engine — Advance Schedule (server-side mirror).

Python port of adjuster-portal-ui/src/app/config/advance-schedule.ts.

This is the server-side source of truth for what constitutes a valid
advance request. The dialog enforces the same policy in the browser as
a UX guardrail, but the backend holds the line — mutating only one side
must not open a loophole.

Keep this file and its TypeScript twin in lockstep. A comment at the
top of each should reference the other when either is edited.

Policy
──────
  Tier table:
    $10,000 – $50,000     → $250
    $50,001 – $100,000    → $500
    $100,001 – $200,000   → $1,000
    $200,001 – open       → $1,500

  Weekly cap (Mon–Sun, per member):  $5,000
  Lifetime cap (per member):         $25,000

  Under-minimum policy: an estimate below the lowest tier minimum
  ($10,000) has no tier amount. The dialog surfaces a free-text
  amount input guarded by an admin-override toggle; the backend
  likewise requires `admin_override=True` before issuing.

Backend enforcement applies whenever `claim_id` is provided AND the
claim has a non-null `estimate_amount`:
  - If estimate falls in a tier:
      amount MUST equal tier.amount unless admin_override=True
  - If estimate is below the minimum tier:
      admin_override=True is required for ANY amount

When `claim_id` is None or the claim has no estimate_amount, we fall
back to admin-discretionary: admin_override=True is required.

`admin_override` NEVER bypasses the weekly or lifetime caps — those
are hard limits regardless.
"""

from __future__ import annotations

from decimal import Decimal
from typing import NamedTuple


class AdvanceTier(NamedTuple):
    """A single bracket in the tier table.

    `max` of None means "no upper bound" — the open top bracket.
    An estimate E matches this tier when `min <= E and (max is None or E <= max)`.
    """
    min: Decimal
    max: Decimal | None
    amount: Decimal


# Ordered by increasing min — `compute_tier_amount` relies on this.
ADVANCE_TIERS: tuple[AdvanceTier, ...] = (
    AdvanceTier(Decimal("10000"),  Decimal("50000"),  Decimal("250")),
    AdvanceTier(Decimal("50001"),  Decimal("100000"), Decimal("500")),
    AdvanceTier(Decimal("100001"), Decimal("200000"), Decimal("1000")),
    AdvanceTier(Decimal("200001"), None,              Decimal("1500")),
)

WEEKLY_CAP_PER_MEMBER = Decimal("5000")
LIFETIME_CAP_PER_MEMBER = Decimal("25000")
UNDER_MINIMUM_POLICY = "admin_discretionary"


def compute_tier_amount(estimate: Decimal | None) -> Decimal | None:
    """Return the tier-matched advance amount for the given estimate.

    None return cases:
      - estimate is None
      - estimate <= 0
      - estimate is below the lowest tier minimum (under-minimum policy)

    Callers distinguish "no match" (None) from a real $0 match. $0 is
    never a valid tier amount in the current schedule.
    """
    if estimate is None or estimate <= 0:
        return None
    for tier in ADVANCE_TIERS:
        if estimate >= tier.min and (tier.max is None or estimate <= tier.max):
            return tier.amount
    return None
